types/transaction.go

```diff
@@ -118,7 +118,14 @@ func (t *Transaction) Copy() *Transaction {
 
 // Cost returns gas * gasPrice + value
 func (t *Transaction) Cost() *big.Int {
-	total := new(big.Int).Mul(t.GasPrice, new(big.Int).SetUint64(t.Gas))
+	total := new(big.Int)
+
+	if t.GasFeeCap != nil || t.GasFeeCap.BitLen() > 0 {
+		total = total.Mul(t.GasFeeCap, new(big.Int).SetUint64(t.Gas))
+	} else {
+		total = total.Mul(t.GasPrice, new(big.Int).SetUint64(t.Gas))
+	}
```

Comment:
Probably good to simplify to something like

```suggestion
        var factor uint64
	if t.GasFeeCap != nil || t.GasFeeCap.BitLen() > 0 {
                 factor = t.GasFeeCap
	} else {
		factor = t.GasPrice
	}
	total = total.Mul(factor, new(big.Int).SetUint64(t.Gas))
```

Reply:
Done